lxd/identity/cache.go

```diff
@@ -0,0 +1,151 @@
+package identity
+
+import (
+	"crypto/x509"
+	"fmt"
+	"net/http"
+	"sync"
+
+	"github.com/canonical/lxd/shared"
+	"github.com/canonical/lxd/shared/api"
+)
+
+// Cache represents an thread-safe in-memory cache of the identities in the database.
+type Cache struct {
+	entries map[string]map[string]CacheEntry
+	mu      sync.RWMutex
+}
+
+// CacheEntry represents an identity. The Certificate field is pre-computed for identities with AuthenticationMethod
+// api.AuthenticationMethodTLS.
+type CacheEntry struct {
+	Identifier           string
+	AuthenticationMethod string
+	IdentityType         string
+	Projects             []string
+	Certificate          *x509.Certificate
+}
+
+// Get returns a single CacheEntry by its authentication method and identifier.
+func (c *Cache) Get(authenticationMethod string, identifier string) (*CacheEntry, error) {
+	c.mu.RLock()
+	defer c.mu.RUnlock()
+
+	if c.entries == nil {
+		c.entries = make(map[string]map[string]CacheEntry)
+	}
+
+	entriesByAuthMethod, ok := c.entries[authenticationMethod]
+	if !ok {
+		return nil, api.StatusErrorf(http.StatusNotFound, "Identity %q (%s) not found", identifier, authenticationMethod)
+	}
+
+	entry, ok := entriesByAuthMethod[identifier]
+	if !ok {
+		return nil, api.StatusErrorf(http.StatusNotFound, "Identity %q (%s) not found", identifier, authenticationMethod)
+	}
+
+	return &entry, nil
+}
+
+// GetByType returns a map of identifier to CacheEntry, where all entries have the given identity type.
+func (c *Cache) GetByType(identityType string) map[string]CacheEntry {
+	c.mu.RLock()
+	defer c.mu.RUnlock()
+
+	if c.entries == nil {
+		c.entries = make(map[string]map[string]CacheEntry)
+	}
+
+	var authenticationMethod string
+	switch identityType {
+	case api.IdentityTypeCertificateServer, api.IdentityTypeCertificateMetrics, api.IdentityTypeCertificateClientRestricted, api.IdentityTypeCertificateClientUnrestricted:
+		authenticationMethod = api.AuthenticationMethodTLS
+	default:
+		return nil
+	}
+
+	entriesByAuthMethod, ok := c.entries[authenticationMethod]
+	if !ok {
+		return nil
+	}
+
+	entriesOfType := make(map[string]CacheEntry, len(entriesByAuthMethod))
+	for _, entry := range entriesByAuthMethod {
+		if entry.IdentityType == identityType {
+			entriesOfType[entry.Identifier] = entry
+		}
+	}
+
+	return entriesOfType
+}
+
+// GetByAuthenticationMethod returns a map of identifier to CacheEntry, where all entries have the given authentication
+// method.
+func (c *Cache) GetByAuthenticationMethod(authenticationMethod string) map[string]CacheEntry {
+	c.mu.RLock()
+	defer c.mu.RUnlock()
+
+	if c.entries == nil {
+		c.entries = make(map[string]map[string]CacheEntry)
+	}
+
+	entriesOfAuthMethod, ok := c.entries[authenticationMethod]
+	if !ok {
+		return nil
+	}
+
+	entriesOfAuthMethodCopy := make(map[string]CacheEntry, len(entriesOfAuthMethod))
```

Comment:
this var is named ...Copy, but the other similar functions don't have vars named ...Copy, is this intentional?

Reply:
When getting by identity type we have to make a new map with just those identities.

When getting by authentication method we could just return `c.entries[authenticationMethod]`, but this wouldn't be thread-safe because a map is a pointer and the caller could read/edit it without a lock.

I'll add a comment to make this clearer.